src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs

```diff
@@ -1025,6 +1066,223 @@ private static string GetFirstLineSubString(string stringToComplete, out bool ha
             return keyValuePairWithCursor;
         }
 
+        // Pulls the variable out of an assignment's LHS expression
+        // Also brings back the innermost type constraint if there is one
+        private static VariableExpressionAst GetVariableFromExpressionAst(
+            ExpressionAst expression,
+            ref Type typeConstraint,
+            ref ValidateSetAttribute setConstraint)
+        {
+            switch (expression)
+            {
+                // $x = ...
+                case VariableExpressionAst variableExpression:
+                    return variableExpression;
+
+                // [type]$x = ...
+                case ConvertExpressionAst convertExpression:
+                    typeConstraint = convertExpression.Type.TypeName.GetReflectionType();
+                    return GetVariableFromExpressionAst(convertExpression.Child, ref typeConstraint, ref setConstraint);
+
+                // [attribute()][type]$x = ...
+                case AttributedExpressionAst attributedExpressionAst:
+
+                    try
+                    {
+                        setConstraint = attributedExpressionAst.Attribute.GetAttribute() as ValidateSetAttribute;
+                    }
+                    catch
+                    {
+                        // Do nothing, just prevent fallout from an unsuccessful attribute conversion
+                    }
+
+                    return GetVariableFromExpressionAst(attributedExpressionAst.Child, ref typeConstraint, ref setConstraint);
+
+                // Something else, like `MemberExpressionAst` $a.p = <tab> which isn't currently handled
+                default:
+                    return null;
+            }
+        }
+
+        // Gets any type constraints or validateset constraints on a given variable
+        private static bool TryGetTypeConstraintOnVariable(
+            CompletionContext completionContext,
+            string variableName,
+            out Type typeConstraint,
+            out ValidateSetAttribute setConstraint)
+        {
+            typeConstraint = null;
+            setConstraint = null;
+
+            PSVariable variable = completionContext.ExecutionContext.EngineSessionState.GetVariable(variableName);
```

Comment:
This doesn't feel like the right thing to do for 2 reasons:
1. this will get you the variable defined in parent scope, however, assignment in a child scope will create a new variable in the child scope, so it's not clear if the completion should honor the constrains of the variable from parent scope with the same name.
2. this doesn't work when editing in VS Code, namely the variable is declared in the script that is being edited but nothing actually runs yet.

Reply:
Ultimately completions are a best guess, especially in PowerShell, but in this particular scenario I think we handle:

- When the variable is explicitly constrained in the assignment (based on the AST)
- When the variable has been constrained at runtime (based on the `PSVariable` attributes), which this is the case here

The second case I think you're describing is:

```powershell
[Animal]$Animal = 'Horse'

$Animal = <tab>
```
where we haven't run this script yet.

Without some way of folding through the script (which I'm not sure we have a full capability for), there's no way of knowing that information. We could add that later with more work.

The other case with the scoping issue feels less common to me than having previously executed some similar script or state and wanting to reassign, either interactively or because of the "execute my script for intellisense" habit in VSCode.

Still valid that you might get inappropriate completions for arbitrary variables based on scope, but this will always work for preference variables and in common use scenarios

Reply:
> Without some way of folding through the script (which I'm not sure we have a full capability for), there's no way of knowing that information. We could add that later with more work.

Try `[System.IO.FileInfo]$bbb = "fff"; $bbb.<tab>`, and you will see completion results for members of `FileInfo` type. We have inference code for tab completion, and it looks to me we are already able to get the type constrains of the variable in the script.

As for the scope concern, I think the current way of dealing it is especially useful for the preference variables because when using those variables in nested scope, they don't have the type constraint as in the built-in ones in the global scope.

Reply:
I think for this, we can leave as-is and get feedback